File: 2022/03.ts
// https://adventofcode.com/2022/day/3

function getPriority(char: string): number {
  const charCode = char.charCodeAt(0);
  return charCode >= 97 ? charCode - 96 : charCode - 38;
}

export function part1(input: string[]): number {
  return input
    .map(contents => {
      const half1 = contents.substring(0, contents.length / 2);
      const half2 = contents.substring(contents.length / 2);

      let matchingChar: string;

      for (let i = 0; i < half2.length; i++) {
        const half2Char = half2[i];

        if (half1.indexOf(half2Char) !== -1) {
          matchingChar = half2Char;
          break;
        }
      }

      return getPriority(matchingChar);
    })
    .reduce((a, b) => a + b);
}

export function part2(input: string[]): number {
  const priorities: number[] = [];

  for (let i = 2; i < input.length; i += 3) {
    const contents = input[i];

    let matchingChar: string;

    for (let j = 0; j < contents.length; j++) {
      const thisChar = contents[j];

      if (input[i - 1].indexOf(thisChar) !== -1 && input[i - 2].indexOf(thisChar) !== -1) {
        matchingChar = thisChar;
        break;
      }
    }

    priorities.push(getPriority(matchingChar));
  }

  return priorities.reduce((a, b) => a + b);
}

File: 2022/src/bin/03.rs
// https://adventofcode.com/2022/day/3

fn char_priority(char: char) -> u16 {
    let char_code = char as u16;

    if char_code >= 97 {
        char_code - 96
    } else {
        char_code - 38
    }
}

fn part1(input: &str) -> u16 {
    input
        .lines()
        .map(|line| {
            let half1 = &line[..(line.len() / 2)];
            let half2 = &line[(line.len() / 2)..];
            let char = half1.chars().find(|&char| half2.contains(char));

            char_priority(char.unwrap())
        })
        .sum()
}

fn part2(input: &str) -> u16 {
    input
        .lines()
        .collect::<Vec<&str>>()
        .chunks(3)
        .map(|group| {
            let char = group[0]
                .chars()
                .find(|&char| group[1].contains(char) && group[2].contains(char));

            char_priority(char.unwrap())
        })
        .sum()
}

fn main() {
    let input = include_str!("../../input/03.txt").trim_end();

    println!("part 1: {}", part1(input));
    println!("part 2: {}", part2(input));
}
